File: PolarsVsSpark/spark/bench2.scala
import org.apache.spark.sql.functions._
import org.apache.spark.sql.{DataFrame, SparkSession}

object bench2 {
  def read_csvs(spark: SparkSession, path: String): DataFrame =
    spark.read
      .option("recursiveFileLookup", "true")
      .option("header", "true")
      .csv(path)

  def write_parquets(df: DataFrame, path: String): Unit =
    df.write.format("parquet").mode("overwrite").save(path)

  def calculate_metrics(df: DataFrame): DataFrame =
    df.groupBy(
      col("date"),
      year(col("date")).alias("year"),
      month(col("date")).alias("month"),
      dayofmonth(col("date")).alias("day"),
      col("model")
    ).agg(sum("failure").alias("failures"))

  def main(args: Array[String]): Unit = {
    val t1 = System.nanoTime()

    val read_path = "root/input/*.csv"
    val write_path = "root/parquets/parquets"
    val spark = SparkSession.builder.getOrCreate()

    val raw_data = read_csvs(spark, read_path)
    write_parquets(raw_data, write_path)

    val t2 = System.nanoTime()
    println(
      "Time to run Apache Spark Load csv and save parquet : " + (t2 - t1) / 1e9 + "s"
    )
    spark.stop()
  }
}

File: PolarsVsSpark/polars/bench2.py
import polars as pl
from datetime import datetime


def read_csvs(path: str):
    lazy_df = pl.scan_csv(path)
    return lazy_df


def write_parquets(df, path: str) -> None:
    df.sink_parquet(path, compression='snappy')

def main():
    t1 = datetime.now()

    read_path = 'root/input/*.csv'
    write_path = 'root/parquets/parquets2'

    df = read_csvs(read_path)
    write_parquets(df, write_path)

    t2 = datetime.now()
    print(f"Time to run polars pipeline : {t2 - t1}")


if __name__ == "__main__":
    main()
